drivers/lsm6dsox/lsm6dsox.py

```diff
@@ -88,10 +97,15 @@ def __init__(
         accel_scale: (+/-2g, +/-4g, +/-8g, +-16g)
         ucf: MLC program to load.
         """
-        self.i2c = i2c
+        self.bus = bus
+        self.cs_pin = cs_pin
         self.address = address
 
+        if cs_pin is None and isinstance(self.bus, SPI):
```

Comment:
I suggest to check for attributes to determine the bus kind.  Eg:
```python
self._use_i2c = hasattr(self.bus, "readfrom_mem")
if cs_pin is None and not self._use_i2c:
    raise ValueError(...)
```

Reply:
It's not likely, but what if another bus/obj has that attr ? I could just check if `cs_pin` is set or not?

Reply:
Using `readfrom_mem` for now.

Reply:
The idea for using hasattr like this is that you can provide any class as the bus, so long as it has the correct methods.  So strictly you want something like:
```python
if hasattr(bus, "read") and hasattr(bus, "write"):
   self._bus_kind = "SPI"
elif hasattr(bus, "readfrom_mem") and hasattr(bus, "writeto_mem"):
    self._bus_kind = "I2C"
```
But I just tried to make it a simpler version of that, and I think it's enough to just check for `readfrom_mem`.  If you require either an I2C-like instance or SPI-like instance, checking for `readfrom_mem` should be enough to distinguish them.

Reply:
But, yes, the simpler alternative is to just check for `cs_pin` being None or not, and use that to decide if it's SPI or I2C.  That might actually be better, then you can just use `if cs_pin: ...` in the register helper methods.

Reply:
I actually went ahead and implemented your first suggestion, after thinking about it the user may forget to set the cs_pin and still want SPI. However, let me know if you want to switch to checking `cs_pin` either way is fine by me.

Reply:
> Using readfrom_mem for now.

I do like the idea of just using `cs_pin`.  That is the simplest option, and means you don't need to raise an exception if `cs_pin` is not given but the bus is detected as SPI.

Reply:
I could add this:

```Python
if hasattr(bus, "read") and hasattr(bus, "write"):
   self._bus_kind = "SPI"
elif hasattr(bus, "readfrom_mem") and hasattr(bus, "writeto_mem"):
    self._bus_kind = "I2C"
else:
    raise ValueError("Unsupported bus")

if self._bus_kind == 'SPI' and self.cs_pin is None:
    raise ValueError("A CS pin must be provided in SPI mode.")
```

Reply:
That's overkill, IMO.

You can just leave it as-is with the `readfrom_mem` used to detect.

Reply:
Okay, just updated the exception message.